### PR TITLE
win-capture: Add compatibility notices

### PR DESCRIPTION
**Pull request by derrod**
### Description

Adds text warnings with optional URLs to window and game capture to notify users about incompatibilities using the new text labels introduced in #5160.

The update mechanism for the compatibility data is the same as used in rtmp-services, meaning that it will fetch updated data from the obsproject website on startup.

**Screenshots:**
Window capture (warning):
2023-02-05_17-26-21_mGUCRU

### Motivation and Context

Provide in-app guidance for users when things don't work in case of known compatibility issues.

### How Has This Been Tested?

Mucked about locally.

### Types of changes

- New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by derrod** [comment]
Based on off-chain feedback translation is a potential issue.

My idea would be:
- Add `translation_key` and `title` properties to JSON
- Provide translatable generic messages, like `%title% may require OBS to be run as admin to use Game Capture`
- Provide translatable specified message, e.g. for CSGO and Minecraft
- If the translation string exists, attempt to replace variables, i.e. `%title%` in string => `title` from JSON
- If no translation key exists, e.g. outdated OBS version or new entry where generic messages couldn't be used, use message from JSON instead

**Comment by derrod** [comment]
Rebased on master with the following changes:

- Remove labels (as discussed off-chain yesterday)
- Added explanation of severity labels to schema
- Squashed commits

**Comment by Gol-D-Ace** [comment]
Gave this a try and it looks like if no application is selected it defaults to the CSGO message for me.
Screenshot_20230205-162221-489

**Comment by derrod** [comment]
Yeah I didn't consider the "any" case. I fixed that now (Also added GTA:SA and SAMP entries based on Discord discussion).

**Review by Warchamp7 — APPROVED** [review]

File: plugins/win-capture/data/locale/en-US.ini [hunk 1]
```diff
@@ -39,3 +39,13 @@ GameCapture.Rgb10a2Space="RGB10A2 Color Space"
...
 Mode="Mode"
+
+# Generic compatibility messages
+Compatibility.GameCapture.Admin="%name% may require OBS to be run as admin to use Game Capture."
+Compatibility.GameCapture.Blocked="%name% cannot be captured via Game Capture, use Window Capture or Display Capture instead."
+Compatibility.GameCapture.WrongGPU="If the preview is blank, make sure %name% is running on the same GPU as OBS."
+Compatibility.WindowCapture.BitBlt="%name% may not be capturable using the selected Capture Method (BitBlt)."
+
+# Specific compatibility messages
+Compatibility.Application.CSGO="CSGO may require the <code>--allow_third_party_software</code> launch option to use Game Capture"
```
Comment (RytoEX) [inline plugins/win-capture/data/locale/en-US.ini:50]:
```suggestion
Compatibility.Application.CSGO="CS:GO may require the <code>--allow_third_party_software</code> launch option to use Game Capture"
```

File: plugins/win-capture/data/locale/en-US.ini [hunk 2]
```diff
@@ -39,3 +39,13 @@ GameCapture.Rgb10a2Space="RGB10A2 Color Space"
...
+Compatibility.Application.Minecraft="If you're having issues capturing Minecraft: Java Edition check our troubleshooting guide"
```
Comment (RytoEX) [inline plugins/win-capture/data/locale/en-US.ini:51]: Sentence ending punctuation?

File: plugins/win-capture/data/schema/compatibility-schema-v1.json [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            "minimum": 1,
+            "maximum": 7
+        },
+        "severityLevels": {
+            "type": "integer",
+            "minimum": 0,
+            "maximum": 2
+        }
+    }
+}
```
Comment (RytoEX) [inline plugins/win-capture/data/schema/compatibility-schema-v1.json:99]: End with newline.

File: plugins/win-capture/game-capture.c [hunk 4]
```diff
@@ -2247,8 +2249,38 @@ static bool mode_callback(obs_properties_t *ppts, obs_property_t *p,
...
+	bool modified = ms_check_window_property_setting(
+		ppts, p, settings, SETTING_CAPTURE_WINDOW, 1);
+
+	if (obs_data_get_bool(settings, SETTING_ANY_FULLSCREEN))
+		return modified;
+
+	const char *window =
+		obs_data_get_string(settings, SETTING_CAPTURE_WINDOW);
+
+	char *class, *exe, *title;
```
Comment (RytoEX) [inline plugins/win-capture/game-capture.c:2261]: Declare these on separate lines.

File: plugins/win-capture/plugin-main.c [hunk 5]
```diff
@@ -75,16 +107,30 @@ bool obs_module_load(void)
...
+
+		if (local_dir) {
+			update_info = update_info_create(
+				WIN_CAPTURE_LOG_STRING, WIN_CAPTURE_VER_STRING,
+				update_url, local_dir, config_dir,
+				confirm_compat_file, NULL);
+		}
 	}
+	bfree(local_dir);
+	bfree(config_dir);
```
Comment (RytoEX) [inline plugins/win-capture/plugin-main.c:133]: Free in reverse order of declaration.

**Review by RytoEX — CHANGES_REQUESTED** [review]

File: .github/workflows/compat-json.yml [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      - "plugins/win-capture/data/package.json"
+  pull_request:
+    paths:
+      - "plugins/win-capture/data/compatibility.json"
+      - "plugins/win-capture/data/package.json"
+
+jobs:
+  schema:
+    name: Schema
+    runs-on: [ubuntu-20.04]
```
[the remark this answers is not in the document]
Comment (derrod) [reply .github/workflows/compat-json.yml:16]: No specific reason, this was just copied 1:1 from the rtmp-service schema checker.

File: plugins/win-capture/data/compatibility.json [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+            "translation_key": "Compatibility.GameCapture.Admin",
+            "severity": 1,
+            "executable": "VALORANT-Win64-Shipping.exe",
+            "window_class": "",
+            "window_title": "",
+            "match_flags": 1,
+            "game_capture": true,
+            "window_capture": false,
+            "window_capture_wgc": false,
+            "message": "Valorant may require OBS to be run as admin to use Game Capture.",
```
[the remark this answers is not in the document]
Comment (derrod) [reply plugins/win-capture/data/compatibility.json:72]: I do not have that game installed, I just added entries from https://obsproject.com/kb/game-capture-troubleshooting

File: plugins/win-capture/plugin-main.c [hunk 8]
```diff
@@ -1,8 +1,18 @@
...
 #include <util/windows/win-version.h>
 #include <util/platform.h>
 
+#include <file-updater/file-updater.h>
+
+#include "compat-helpers.h"
+#include "compat-format-ver.h"
+#include "compat-config.h"
+
+#define WIN_CAPTURE_LOG_STRING "[win-capture plugin] "
```
[the remark this answers is not in the document]
Comment (derrod) [reply plugins/win-capture/plugin-main.c:13]: This is just a straight copy from `rtmp-services`:

https://github.com/obsproject/obs-studio/blob/master/plugins/rtmp-services/rtmp-services-main.c#L21-L22

An example line would be
`20:47:45.242: [win-capture plugin] Remote update of URL "https://obsproject.com/obs2_update/win-capture/v1/package.json" failed: The requested URL returned error: 404`

[the remark this answers is not in the document]
Comment (RytoEX) [reply plugins/win-capture/plugin-main.c:13]: Weird. If it's only used in that one place and is based on an existing pattern, then I'm not as concerned.

**Comment by derrod** [comment]
Based on Flaeri's feedback:
- Removed Valorant (seems to be no longer reuqired, possibly never was)
- Added generic Electron entries to BitBlt Window Capture and Game Capture
  + Window Capture will default to WGC for Electron apps, so only a "notice" severity
  + For Game Capture this affects titles such as Vampire Survivors that are built using Electron. See https://steamdb.info/tech/SDK/Greenworks/ for a list of many affected titles

**Review by RytoEX — APPROVED** [review]
Looks okay to me. I think this is a good first attempt at trying to give users some assistance to resolve issues on their own without having to come to our support channels.

**Pull request merged**